regression: destructor of `[MaybeUninit<T>; N]` cannot be evaluated at compile-time

### Code

I tried this code:

```rust
pub const fn f<T, const N: usize>(_: [std::mem::MaybeUninit<T>; N]) {}
```

I expected it to compile successfully. Instead, this happened:

```
error[E0493]: destructor of `[MaybeUninit<T>; N]` cannot be evaluated at compile-time
 --> x.rs:1:35
  |
1 | pub const fn f<T, const N: usize>(_: [std::mem::MaybeUninit<T>; N]) {}
  |                                   ^                                  - value is dropped here
  |                                   |
  |                                   the destructor for this type cannot be evaluated in constant functions
```


### Version it worked on

It most recently worked on 1.72

### Version with regression

Regressed in aafd75a9c510b0e91746b891eb4ebade43899af5 cc @fee1-dead @oli-obk

Originally reported in https://github.com/rust-lang/rust/issues/115403#issuecomment-1701000117 cc @itsxaos

@rustbot modify labels: +regression-from-stable-to-beta -regression-untriaged

--- a/compiler/rustc_ty_utils/src/needs_drop.rs
+++ b/compiler/rustc_ty_utils/src/needs_drop.rs
@@ -19,11 +19,30 @@ fn needs_drop_raw<'tcx>(tcx: TyCtxt<'tcx>, query: ty::ParamEnvAnd<'tcx, Ty<'tcx>
     // needs drop.
     let adt_has_dtor =
         |adt_def: ty::AdtDef<'tcx>| adt_def.destructor(tcx).map(|_| DtorType::Significant);
-    let res =
-        drop_tys_helper(tcx, query.value, query.param_env, adt_has_dtor, false).next().is_some();
+    let res = drop_tys_helper(tcx, query.value, query.param_env, adt_has_dtor, false)
+        .filter(filter_array_elements(tcx, query.param_env))
+        .next()
+        .is_some();
 
     debug!("needs_drop_raw({:?}) = {:?}", query, res);
     res
+}
+
+/// HACK: in order to not mistakenly assume that `[PhantomData<T>; N]` requires drop glue
+/// we check the element type for drop glue. The correct fix would be looking at the
+/// entirety of the code around `needs_drop_components` and this file and come up with
+/// logic that is easier to follow while not repeating any checks that may thus diverge.
+fn filter_array_elements<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    param_env: ty::ParamEnv<'tcx>,
+) -> impl Fn(&Result<Ty<'tcx>, AlwaysRequiresDrop>) -> bool {
+    move |ty| match ty {
+        Ok(ty) => match *ty.kind() {
+            ty::Array(elem, _) => tcx.needs_drop_raw(param_env.and(elem)),
+            _ => true,
+        },
+        Err(AlwaysRequiresDrop) => true,
+    }
 }
 
 fn has_significant_drop_raw<'tcx>(
@@ -37,6 +56,7 @@ fn has_significant_drop_raw<'tcx>(
         adt_consider_insignificant_dtor(tcx),
         true,
     )
+    .filter(filter_array_elements(tcx, query.param_env))
     .next()
     .is_some();
     debug!("has_significant_drop_raw({:?}) = {:?}", query, res);

--- a/tests/ui/consts/drop-maybe_uninit.rs
+++ b/tests/ui/consts/drop-maybe_uninit.rs
@@ -1,0 +1,17 @@
+// build-pass
+
+pub const fn f<T, const N: usize>(_: [std::mem::MaybeUninit<T>; N]) {}
+
+pub struct Blubb<T>(*const T);
+
+pub const fn g<T, const N: usize>(_: [Blubb<T>; N]) {}
+
+pub struct Blorb<const N: usize>([String; N]);
+
+pub const fn h(_: Blorb<0>) {}
+
+pub struct Wrap(Blorb<0>);
+
+pub const fn i(_: Wrap) {}
+
+fn main() {}

